### PR TITLE
Fix /emergeblocks crashing in debug builds

### LINKED ISSUE #5441: /emergeblocks crashes
If I run /emergeblocks I get this error message after a while

`minetest: /build/minetest-scNidi/minetest-201703201847/src/emerge.cpp:391: bool EmergeManager::popBlockEmergeData(v3s16, BlockEmergeData*): Zusicherung »count_peer != 0« nicht erfüllt.`

Example command:

`/emergeblocks (-1000, -20, -1000) (1000, 47, 1000)`

I'm using Ubuntu 16.04 64 Bit with the latest build from the launchpad PPA. Tested with a new v6 and v7 map without mods and sqlite3, leveldb and dummy database backend

### PR DESCRIPTION
**Pull request by Wuzzy2**
Fixes #5441.

The reason for the bug was an u16 overflow, thus failing the assert. This only happened in Debug build but not in Release builds, which is why it didn't happen to paramat.

The bugfix is based on what the ghost account (=deleted account) wrote in the issue. And I added `m_qlimit_*` to the u32 change.

While I was at it, I also added lower and upper bounds to the `emergequeue_*` settings.

Please test and verify to make sure Minetest doesn't explode or something. ;)

### CODE REVIEW

**Comment by SmallJoker** [comment]
Alternatively:
https://github.com/minetest/minetest/blob/f4d8cc0f0bf33a998aa0d6d95de4b34f69fb31db/src/emerge.cpp#L440-L442

Change to `(u32)count_peer * 2 >= (u32)m_qlimit_total`. This should fix the problem too.

**Comment by Wuzzy2** [comment]
@SmallJoker: Nope, this code crashes like in #5441.

File: builtin/settingtypes.txt [hunk 1]
```diff
@@ -2218,15 +2218,15 @@ chunksize (Chunk size) int 5
 enable_mapgen_debug_info (Mapgen debug) bool false
 
 #    Maximum number of blocks that can be queued for loading.
-emergequeue_limit_total (Absolute limit of queued blocks to emerge) int 1024
+emergequeue_limit_total (Absolute limit of queued blocks to emerge) int 1024 1 4294967295
```
Comment (sfan5) [inline builtin/settingtypes.txt:2221]: Although this is the theoretical limit I suggest putting `1000000` instead.

File: src/emerge.cpp [hunk 2]
```diff
@@ -468,7 +468,7 @@ bool EmergeManager::pushBlockEmergeData(
 bool EmergeManager::popBlockEmergeData(v3s16 pos, BlockEmergeData *bedata)
 {
 	std::map<v3s16, BlockEmergeData>::iterator it;
-	std::unordered_map<u16, u16>::iterator it2;
+	std::unordered_map<u16, u32>::iterator it2;
```
Comment (sfan5) [inline src/emerge.cpp:471]: If you're editing this anyway please move `it` and `it2` inline with `auto` as type

**Comment by Wuzzy2** [comment]
Alrighy then. This PR is ready again.

**Review by sfan5 — APPROVED** [review]
LGTM

File: src/emerge.cpp [hunk 3]
```diff
@@ -165,11 +165,11 @@ EmergeManager::EmergeManager(Server *server)
...
+	m_qlimit_total = g_settings->getU32("emergequeue_limit_total");
 	// FIXME: these fallback values are probably not good
-	if (!g_settings->getU16NoEx("emergequeue_limit_diskonly", m_qlimit_diskonly))
+	if (!g_settings->getU32NoEx("emergequeue_limit_diskonly", m_qlimit_diskonly))
 		m_qlimit_diskonly = nthreads * 5 + 1;
-	if (!g_settings->getU16NoEx("emergequeue_limit_generate", m_qlimit_generate))
+	if (!g_settings->getU32NoEx("emergequeue_limit_generate", m_qlimit_generate))
 		m_qlimit_generate = nthreads + 1;
 
 	// don't trust user input for something very important like this
```
Comment (SmallJoker) [inline src/emerge.cpp:175]:
```C++
m_qlimit_total = rangelim(m_qlimit_total, 1, 1000000);
m_qlimit_diskonly = rangelim(m_qlimit_diskonly, 1, 1000000);
m_qlimit_generate = rangelim(m_qlimit_generate, 1, 1000000);
```
if you're already specifying limits, these should also be enforced.

**Comment by Wuzzy2** [comment]
@SmallJoker: Done.

**Review by SmallJoker — APPROVED** [review]
Thanks for the changes

**Pull request merged**